src/Razor/Razor.Runtime/src/Runtime/RazorRazorBasePage.cs

```diff
@@ -0,0 +1,468 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Html;
+using Microsoft.AspNetCore.Razor.Runtime.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Razor
+{
+    /// <summary>
+    /// Represents properties and methods that are needed in order to render a view that uses Razor syntax.
+    /// </summary>
+    public abstract class RazorRazorPageBase
+    {
+        private readonly Stack<TextWriter> _textWriterStack = new Stack<TextWriter>();
+        private StringWriter _valueBuffer;
+        private AttributeInfo _attributeInfo;
+        private TagHelperAttributeInfo _tagHelperAttributeInfo;
+
+        public string Layout { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="TextWriter"/> that the page is writing output to.
+        /// </summary>
+        public TextWriter Output { get; set; }
+
+        /// <inheritdoc />
+        public string Path { get; set; }
+
+        /// <inheritdoc />
+        // REVIEW: using Func<Task> instead of moving RenderAsyncDelegate around, which is a breaking change
+        public IDictionary<string, Func<Task>> SectionWriters { get; } =
+            new Dictionary<string, Func<Task>>(StringComparer.OrdinalIgnoreCase);
+
+        /// <inheritdoc />
+        public bool IsLayoutBeingRendered { get; set; }
+
+        /// <inheritdoc />
+        public IHtmlContent BodyContent { get; set; }
+
+        /// <inheritdoc />
+        public IDictionary<string, Func<Task>> PreviousSectionWriters { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="System.Text.Encodings.Web.HtmlEncoder"/> to use when this <see cref="RazorRazorPageBase"/>
+        /// handles non-<see cref="IHtmlContent"/> C# expressions.
+        /// </summary>
+        public HtmlEncoder HtmlEncoder { get; set; } = HtmlEncoder.Default; // REVIEW: how should we [RazorInject] this in a non-breaking way?
```

Comment:
I was thinking we could do the following for `[RazorInject]`.

Introduce a new type `Microsoft.AspNetCore.Razor.RazorInjectAttribute`.Mark the old one obsolete, but make MVC still honor it. We might want to come up with another name for the new one so they don't clash. As we we can still load old types and MVC still supports the old attribute we are OK.